org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/Annotation.java

```diff
@@ -1300,7 +1306,13 @@ static AnnotationTargetAllowed isAnnotationTargetAllowed(Annotation annotation,
 
 		long metaTagBits = annotationType.getAnnotationTagBits(); // could be forward reference
 		if ((metaTagBits & TagBits.AnnotationTargetMASK) == 0) {
-			// does not specify any target restriction - all locations are possible
+			/* JLS 9.6.4.1: If an annotation of type java.lang.annotation.Target is not present on the
+			   declaration of an annotation interface A, then A is applicable in all declaration
+			   contexts and in no type contexts.
+			*/
+			if (kind == Binding.TYPE_USE) {
+				scope.problemReporter().explitAnnotationTargetRequired(annotation);
+			}
 			return AnnotationTargetAllowed.YES;
```

Comment:
> Answering YES after detecting an `explitAnnotationTargetRequired` error looks funny, but I can see that this has been inherited from 10 years ago. Would you see value in inventing an answer `AnnotationTargetAllowed.TYPE_USE_CONTEXT_LACKING_TARGET` and letting the two callers of `isAnnotationTargetAllowed` report the issue along with similar issues? Would this change whether or not we report follow-up errors (like when `checkAnnotationTarget` resets `tagBitsToRevert`)?

I opted for the name `AnnotationTargetAllowed.NO_DUE_TO_LACKING_TARGET` and implemented the change suggested above. 

It results in better error messages in the case of repeating annotations without explicit target being applied to TYPE_USE locations. See org.eclipse.jdt.core.tests.compiler.regression.RepeatableAnnotationTest.testRepeatingAnnotationsWithoutTarget()

Rather than complain 

```
----------\n
1. ERROR in X.java (at line 1)\n
	public class X<@Foo @Foo T> extends @Foo @Foo Object {\n
	                                    ^^^^\n
Annotation types that do not specify explicit target element types cannot be applied here\n
----------\n
2. ERROR in X.java (at line 1)\n
	public class X<@Foo @Foo T> extends @Foo @Foo Object {\n
	                                    ^^^^\n
Annotation types that do not specify explicit target element types cannot be applied here\n
----------\n
3. ERROR in X.java (at line 1)\n
	public class X<@Foo @Foo T> extends @Foo @Foo Object {\n
	                                         ^^^^\n
Annotation types that do not specify explicit target element types cannot be applied here\n
----------\n
```

we now complain:

```
----------\n
1. ERROR in X.java (at line 1)\n
	public class X<@Foo @Foo T> extends @Foo @Foo Object {\n
	                                    ^^^^\n
Annotation types that do not specify explicit target element types cannot be applied here\n
----------\n
2. ERROR in X.java (at line 1)\n
	public class X<@Foo @Foo T> extends @Foo @Foo Object {\n
	                                    ^^^^\n
The annotation @Foo cannot be repeated at this location since its container annotation type @FooContainer is disallowed at this location\n
----------\n
3. ERROR in X.java (at line 1)\n
	public class X<@Foo @Foo T> extends @Foo @Foo Object {\n
	                                         ^^^^\n
Annotation types that do not specify explicit target element types cannot be applied here\n
----------\n
```

As for `tagBitsToRevert` I think we should see what the effect is and take suitable follow up action. Returning Yes from there is bogus - yes we have done that for 10 years. I'll initiate a test run and see what happens